// @marinade.finance/marinade-ts-sdk/dist/src/marinade-native-stake.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.callRebalanceHint = exports.getPrepareNativeUnstakeSOLIx = exports.getRefNativeStakeAccountTx = exports.getAuthNativeStakeAccountIx = exports.getRefNativeStakeSOLTx = exports.getAuthNativeStakeSOLIx = void 0;
const web3_js_1 = require("@solana/web3.js");
const native_staking_sdk_1 = require("@marinade.finance/native-staking-sdk");
/**
 * Retrieves the instruction for staking SOL with Marinade Native.
 *
 * @param {PublicKey} userPublicKey - The PublicKey associated with the user.
 * @param {BN} amountToStake - The amount to be staked, in lamports.
 * @returns {AuthStakeSOLIxResponse} - The instructions to include in the transaction for staking in Marinade Native, along with the new stake keypair.
 */
function getAuthNativeStakeSOLIx(userPublicKey, amountToStake) {
    const nativeSdk = new native_staking_sdk_1.NativeStakingSDK();
    return nativeSdk.buildCreateAuthorizedStakeInstructions(userPublicKey, amountToStake);
}
exports.getAuthNativeStakeSOLIx = getAuthNativeStakeSOLIx;
/**
 * Retrieves the transaction for staking SOL with Marinade Native using a referral code.
 *
 * @param {PublicKey} userPublicKey - The PublicKey associated with the user.
 * @param {BN} amountToStake - The amount to be staked, specified in lamports.
 * @param {string} mNativeRefCode - The Marinade Native referral code.
 * @returns {Promise<VersionedTransaction>} - A promise that resolves to the transaction for staking SOL with Marinade Native using the referral code.
 */
function getRefNativeStakeSOLTx(userPublicKey, amountToStake, mNativeRefCode) {
    return __awaiter(this, void 0, void 0, function* () {
        const response = yield fetch(`https://native-staking-referral.marinade.finance/v1/tx/deposit-sol?amount=${amountToStake.toString()}&code=${mNativeRefCode}&user=${userPublicKey.toString()}`);
        const result = yield response.json();
        const txBuffer = Buffer.from(result.serializedTx, 'base64');
        return web3_js_1.VersionedTransaction.deserialize(txBuffer);
    });
}
exports.getRefNativeStakeSOLTx = getRefNativeStakeSOLTx;
/**
 * Retrieves the instruction for staking a Stake Account with Marinade Native.
 *
 * @param {PublicKey} userPublicKey - The PublicKey associated with the user.
 * @param {PublicKey[]} stakeAccounts - The stake accounts to be used with Marinade Native.
 * @returns {TransactionInstruction} - The instruction to include in the transaction for staking the Stake Account with Marinade Native.
 */
function getAuthNativeStakeAccountIx(userPublicKey, stakeAccounts) {
    const nativeSdk = new native_staking_sdk_1.NativeStakingSDK();
    return nativeSdk.buildAuthorizeInstructions(userPublicKey, stakeAccounts);
}
exports.getAuthNativeStakeAccountIx = getAuthNativeStakeAccountIx;
/**
 * Retrieves the transaction for using a Stake Account with Marinade Native via a referral code.
 *
 * @param {PublicKey} userPublicKey - The PublicKey associated with the user.
 * @param {PublicKey} stakeAccountAddress - The address of the stake account intended to use with Marinade Native.
 * @param {string} mNativeRefCode - The Marinade Native referral code.
 * @returns {Promise<VersionedTransaction>} - A promise resolving to the transaction required for using the Stake Account with Marinade Native via the given referral code.
 */
function getRefNativeStakeAccountTx(userPublicKey, stakeAccountAddress, mNativeRefCode) {
    return __awaiter(this, void 0, void 0, function* () {
        const response = yield fetch(`https://native-staking-referral.marinade.finance/v1/tx/deposit-stake-account?stake=${stakeAccountAddress.toString()}&code=${mNativeRefCode}&user=${userPublicKey.toString()}`);
        const result = yield response.json();
        const txBuffer = Buffer.from(result.serializedTx, 'base64');
        return web3_js_1.VersionedTransaction.deserialize(txBuffer);
    });
}
exports.getRefNativeStakeAccountTx = getRefNativeStakeAccountTx;
/**
 * Retrieves the instruction for unstaking from Marinade Native.
 *
 * @param {PublicKey} userPublicKey - The PublicKey associated with the user.
 * @param {BN} amountToUnstake - The amount to be unstaked, specified in lamports.
 * @returns {UnstakeSOLIxResponse} - The instructions for paying the unstaking fee and an optional event that can be called after the fee is paid to expedite the unstaking process (calling this event is optional but enhances the user experience).
 */
function getPrepareNativeUnstakeSOLIx(userPublicKey, amountToUnstake) {
    return __awaiter(this, void 0, void 0, function* () {
        const nativeSdk = new native_staking_sdk_1.NativeStakingSDK();
        return yield nativeSdk.initPrepareForRevoke(userPublicKey, amountToUnstake);
    });
}
exports.getPrepareNativeUnstakeSOLIx = getPrepareNativeUnstakeSOLIx;
/**
 * Invokes the bot to rebalance the user's funds. This is particularly useful immediately following a deposit into Marinade Native, although it is not mandatory.
 *
 * @param {PublicKey} userPublicKey - The PublicKey associated with the user.
 */
function callRebalanceHint(userPublicKey) {
    return __awaiter(this, void 0, void 0, function* () {
        const nativeSdk = new native_staking_sdk_1.NativeStakingSDK();
        return yield nativeSdk.callRebalanceHint(userPublicKey);
    });
}
exports.callRebalanceHint = callRebalanceHint;
