// @marinade.finance/directed-stake-sdk/src/api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.findVoteRecords = exports.root = exports.voteRecord = void 0;
async function voteRecord({ sdk, address, }) {
    return sdk.program.account.voteRecord.fetch(address);
}
exports.voteRecord = voteRecord;
async function root({ sdk, address, }) {
    return sdk.program.account.root.fetch(address);
}
exports.root = root;
async function findVoteRecords({ sdk, root, owner, target, }) {
    const filters = [];
    if (root) {
        filters.push({
            memcmp: {
                bytes: root.toBase58(),
                offset: 8,
            },
        });
    }
    if (owner) {
        filters.push({
            memcmp: {
                bytes: owner.toBase58(),
                offset: 40,
            },
        });
    }
    if (target) {
        filters.push({
            memcmp: {
                bytes: target.toBase58(),
                offset: 72,
            },
        });
    }
    return await sdk.program.account.voteRecord.all(filters);
}
exports.findVoteRecords = findVoteRecords;
